File: day_13/13.js
/** @param {string} moves
 * @returns {true|[number, number]} Return true if robot returns or position
 */
function isRobotBack(moves) {
  // Code here
  const robot = {
    x: 0,
    y: 0,
  };
  const lastMove = {
    L: false,
    R: false,
    U: false,
    D: false,
  };
  const movesFn = {
    L: (index) => {
      robot.x -= 1;
      lastMove.L = true;
      return index + 1;
    },
    R: (index) => {
      robot.x += 1;
      lastMove.R = true;
      return index + 1;
    },
    U: (index) => {
      robot.y += 1;
      lastMove.U = true;
      return index + 1;
    },
    D: (index) => {
      robot.y -= 1;
      lastMove.D = true;
      return index + 1;
    },
    "*": (index) => {
      const nextMove = moves[index + 1];
      movesFn[nextMove](0);
      movesFn[nextMove](0);
      return index + 2;
    },
    "?": (index) => {
      const nextMove = moves[index + 1];
      if (!lastMove[nextMove]) {
        movesFn[nextMove](index);
      }
      return index + 2;
    },
  };
  let index = 0;
  moves = moves.replaceAll("!U", "D");
  moves = moves.replaceAll("!D", "U");
  moves = moves.replaceAll("!L", "R");
  moves = moves.replaceAll("!R", "L");
  while (index < moves.length) {
    const element = moves[index];
    index = movesFn[element](index);
  }
  return (robot.x === 0 && robot.y === 0) || Object.values(robot);
}

console.log("R", isRobotBack("R"));
console.log("RL", isRobotBack("RL"));
console.log("RLUD", isRobotBack("RLUD"));
console.log("*RU", isRobotBack("*RU"));
console.log("R*U", isRobotBack("R*U"));
console.log("LLL!R", isRobotBack("LLL!R"));
console.log("R?R", isRobotBack("R?R"));
console.log("U?D", isRobotBack("U?D"));
console.log("R!L", isRobotBack("R!L"));
console.log("U!D", isRobotBack("U!D"));
console.log("R?L", isRobotBack("R?L"));
console.log("U?U", isRobotBack("U?U"));
console.log("*U?U", isRobotBack("*U?U"));
console.log("U?D?U", isRobotBack("U?D?U"));

console.log("R!U?U", isRobotBack("R!U?U"));
console.log("UU!U?D", isRobotBack("UU!U?D"));

File: day_13/13.py
def is_robot_back(moves: str) -> bool | list[int]:
    robot = {"x": 0, "y": 0}
    last_move = {"L": False, "R": False, "U": False, "D": False}

    def move_left(index):
        robot["x"] -= 1
        last_move["L"] = True
        return index + 1

    def move_right(index):
        robot["x"] += 1
        last_move["R"] = True
        return index + 1

    def move_up(index):
        robot["y"] += 1
        last_move["U"] = True
        return index + 1

    def move_down(index):
        robot["y"] -= 1
        last_move["D"] = True
        return index + 1

    def move_star(index):
        next_move = moves[index + 1]
        moves_fn[next_move](index)
        moves_fn[next_move](index)
        return index + 2

    def move_question(index):
        next_move = moves[index + 1]
        if not last_move[next_move]:
            moves_fn[next_move](index)
        return index + 2

    moves_fn = {
        "L": move_left,
        "R": move_right,
        "U": move_up,
        "D": move_down,
        "*": move_star,
        "?": move_question,
    }
    moves = moves.replace("!U", "D")
    moves = moves.replace("!D", "U")
    moves = moves.replace("!L", "R")
    moves = moves.replace("!R", "L")
    index = 0
    while index < len(moves):
        move = moves[index]
        index = moves_fn[move](index)

    return True if robot["x"] == 0 and robot["y"] == 0 else [robot["x"], robot["y"]]


print(is_robot_back("R"))
print(is_robot_back("RL"))
print(is_robot_back("RLUD"))
print(is_robot_back("*RU"))
print(is_robot_back("R*U"))
print(is_robot_back("LLL!R"))
print(is_robot_back("R?R"))
print(is_robot_back("U?D"))
print(is_robot_back("R!L"))
print(is_robot_back("U!D"))
print(is_robot_back("R?L"))
print(is_robot_back("U?U"))
print(is_robot_back("*U?U"))
print(is_robot_back("U?D?U"))
print(is_robot_back("R!U?U"))
print(is_robot_back("UU!U?D"))
